csharp/src/Ice/Ice1BinaryConnection.cs

```diff
@@ -0,0 +1,496 @@
+                    s.Append(' ');
+                    s.Append(Endpoint.TransportName);
+                    s.Append(" connection\n");
+                    s.Append(ToString());
+                }
+                Endpoint.Communicator.Logger.Trace(Endpoint.Communicator.TraceLevels.NetworkCategory,
+                                                    s.ToString());
+            }
+        }
+
+        public async ValueTask<(int StreamId, object? Frame, bool Fin)> ReceiveAsync(CancellationToken cancel)
+        {
+            while (true)
+            {
+                int requestId = 0;
+                object? frame = null;
+                Task<ArraySegment<byte>>? task = null;
+                ValueTask<ArraySegment<byte>> receiveTask = PerformReceiveFrameAsync();
+                if (receiveTask.IsCompletedSuccessfully)
+                {
+                    _receiveTask = Task.CompletedTask;
+                    (requestId, frame) = ParseFrame(receiveTask.Result);
+                }
+                else
+                {
+                    _receiveTask = task = receiveTask.AsTask();
+                }
+
+                if (task != null)
+                {
+                    (requestId, frame) = ParseFrame(await task.ConfigureAwait(false));
+                }
+
+                if (frame != null)
+                {
+                    return (StreamId: requestId, Frame: frame, Fin: requestId == 0 || frame is IncomingResponseFrame);
+                }
+            }
+        }
+
+        public int NewStream(bool bidirectional) => bidirectional ? ++_nextStreamId : 0;
+
+        public ValueTask ResetAsync(int streamId) =>
+            throw new NotSupportedException("ice1 transports don't support stream reset");
+
+        public async ValueTask SendAsync(int streamId, object frame, bool fin, CancellationToken cancel) =>
+            await SendFrameAsync(streamId, frame, cancel);
+
+        public override string ToString() => Transceiver.ToString()!;
+
+        internal Ice1BinaryConnection(ITransceiver transceiver, Endpoint endpoint, ObjectAdapter? adapter)
+        {
+            Transceiver = transceiver;
+            Endpoint = endpoint;
+
+            _incoming = adapter != null;
+            _frameSizeMax = adapter?.FrameSizeMax ?? Endpoint.Communicator.FrameSizeMax;
+            _warn = Endpoint.Communicator.GetPropertyAsBool("Ice.Warn.Connections") ?? false;
+            _warnUdp = Endpoint.Communicator.GetPropertyAsBool("Ice.Warn.Datagrams") ?? false;
+            _compressionLevel = Endpoint.Communicator.GetPropertyAsInt("Ice.Compression.Level") ?? 1;
+            _sentCallback = _receivedCallback = _ => {};
+            _heartbeatCallback = () => {};
+
+            if (_compressionLevel < 1)
+            {
+                _compressionLevel = 1;
+            }
+            else if (_compressionLevel > 9)
+            {
+                _compressionLevel = 9;
+            }
```

Comment:
We can throw invalid configuration exception instead

Reply:
This would have to be done on communicator initialization instead and assert here instead. Not really related to the PR so if we want to change this, best is to add a bug.

Reply:
Right, this can be done in a separate PR, maybe we should have an internal property `CompressionLevel` initialized during communicator creation and used here.